docs/design-proposals/2022-04 TraceQL Concepts.md

```diff
@@ -0,0 +1,203 @@
+---
+Authors: Yuri Shkuro (@yurishkuro), Tom Wilkie (@tomwilkie), Cyril Tovena (@cyriltovena), Martin Disibio (@mdisibio), Joe Elliott (@joe-elliott)
+Created: 2021 October - 2022 March
+Last updated: 2022-04-12
+---
+
+# TraceQL Concepts
+
+## Summary
+
+This design document describes a language for selecting traces that Tempo will implement. This language is currently only focused on trace selection. It will be extended to derive metrics from traces in the future.
+
+This document is **not** meant to be a complete language specification. Rather it is meant to invite the community to comment on the basic structure and concepts of the language.
+
+### Table of Contents
+
+- [Capabilities](#capabilities)
+- [Structure](#structure)
+- [Selecting spans](#selecting-spans)
+  - [Intrinsic fields](#intrinsic-fields)
+  - [Attribute fields](#attribute-fields)
+  - [Field expressions](#field-expressions)
+- [Combining Spansets](#combining-spansets)
+  - [Structural Operators](#structural-operators)
+- [Aggregators](#aggregators)
+- [Pipelining](#pipelining)
+- [Grouping](#grouping)
+- [Examples](#examples)
+  - [Shorthand](#shorthand)
+
+## Capabilities
+
+A TraceQL query can select traces based on 
+
+- span attributes, timing and duration
+- structural relationships between spans
+- aggregated data from the spans in a trace
+
+The TraceQL language is designed to use the same syntax and semantics as PromQL and LogQL where possible. The syntax and semantics differ due to the specialized needs of querying traces.
+
+## Structure
+
+A query is an expression that is evaluated on one trace at a time. The query is structured as a set of chained
+expressions (a pipeline). Each expression in the pipeline selects or discards spansets from being included in the results set. E.g.
+
+`{ http.status = 200 } | by(namespace) | count() > 3`
+
+Queries select sets of spans and filter them through a pipeline of aggregators and conditions. If a spanset is produced after evaluation
+on a trace then this spanset (and by extension the trace) is included in the resultset of the query.
+
+## Selecting spans
+
+In TraceQL the curly brackets `{}` always select a set of spans from the current trace. They will commonly be paired with a condition to reduce the spans being passed in.
+
+`{ http.status = 200 }`
+
+This simple query will be evaluated on every span of every trace, one at a time. 
+
+If the trace being evaluated contains no spans with an attribute `http.status` with the value `200` then no spans will be selected and this trace will not appear in our resultset.
+
+If the trace does contain spans with an attribute `http.status` with the value `200` then only those spans will be returned. The trace is reduced to only the set of spans that match the condition inside the `{}`. The resultset will contain only this subset of spans matching the condition.
+
+### Intrinsic fields
```

Comment:
I think span kind is missing from this list. Span kind should be especially useful with the descendant operator.